### PR TITLE
[ALLUXIO-2813] Move promote logic to data path

### PR DESCRIPTION
**Pull request by calvinjia**
https://alluxio.atlassian.net/browse/ALLUXIO-2813

A few things which I think are suboptimal atm

1. Passing a boolean instead of ReadType. If we go with read type we may be able to save more metadata fields, but I think it should be part of a separate refactor.
2. Worker side lock logic. The worker side logic currently assumes locks are held by possibly remote clients which makes locking awkward when we move things to netty (in the same JVM).
3. Threadpool/runnable in short circuit reader. As a developer the code there seems unintuitive.

### CODE REVIEW

File: core/server/worker/src/main/java/alluxio/worker/netty/NettyExecutors.java [hunk 1]
```diff
@@ -49,6 +49,12 @@
           THREAD_STOP_MS, TimeUnit.MILLISECONDS, new SynchronousQueue<Runnable>(),
           ThreadFactoryUtils.build("FilePacketWriterExecutor-%d", true));
 
+  public static final ExecutorService BLOCK_OPEN_EXECUTOR =
+      new ThreadPoolExecutor(THREADS_MIN,
+          Configuration.getInt(PropertyKey.WORKER_NETWORK_NETTY_BLOCK_READER_THREADS_MAX),
+          THREAD_STOP_MS, TimeUnit.MILLISECONDS, new SynchronousQueue<Runnable>(),
+          ThreadFactoryUtils.build("BlockPromoteExecutor-%d", true));
```
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/netty/NettyExecutors.java]: BlockOpenExecutor?

File: core/protobuf/src/proto/dataserver/protocol.proto [hunk 2]
```diff
@@ -88,9 +90,10 @@ message Heartbeat {
 
 // Netty RPCs. Every RPC needs to define a request type and optionally a response type (default to Response).
 
-// next available id: 3
+// next available id: 4
 message LocalBlockOpenRequest {
   optional int64 block_id = 1;
+  optional bool promote = 3;
```
Comment (peisun1115) [inline core/protobuf/src/proto/dataserver/protocol.proto]: s/3/2 ?

File: core/protobuf/src/proto/dataserver/protocol.proto [hunk 3]
```diff
@@ -88,9 +90,10 @@ message Heartbeat {
 
 // Netty RPCs. Every RPC needs to define a request type and optionally a response type (default to Response).
 
-// next available id: 3
+// next available id: 4
```
Comment (peisun1115) [inline core/protobuf/src/proto/dataserver/protocol.proto]: s/4/3

File: core/server/worker/src/main/java/alluxio/worker/netty/NettyExecutors.java (hunk 1 above)
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/netty/NettyExecutors.java]: Updated, thanks.

File: core/protobuf/src/proto/dataserver/protocol.proto (hunk 3 above)
Comment (calvinjia) [inline core/protobuf/src/proto/dataserver/protocol.proto]: Yeah it was strange that it skipped one, but we should be okay to reclaim the space.

File: core/protobuf/src/proto/dataserver/protocol.proto (hunk 2 above)
Comment (calvinjia) [inline core/protobuf/src/proto/dataserver/protocol.proto]: Updated, thanks.

**Comment by peisun1115** [comment]
LGTM

**Pull request merged**